src/Crate/PDO/Http/ServerPool.php

```diff
@@ -144,6 +131,82 @@ public function execute(string $query, array $parameters): CollectionInterface
         );
     }
 
+    /**
+     * {@Inheritdoc}
+     * @throws \GuzzleHttp\Exception\ConnectException
+     */
+    public function execute_bulk(string $query, array $parameters): BulkResponseInterface
+    {
+        $numServers = count($this->availableServers) - 1;
+
+        for ($i = 0; $i <= $numServers; $i++) {
+            // always get the first available server
+            $server = $this->availableServers[0];
+
+            // Move the selected server to the end of the stack
+            $this->availableServers[] = array_shift($this->availableServers);
+
+            try {
+                return $this->send_request($server, $query, $parameters, true);
+
+            } catch (ConnectException $exception) {
+                // Catch it before the BadResponseException but do nothing.
+                continue;
+            } catch (BadResponseException $exception) {
+                $body = (string)$exception->getResponse()->getBody();
+                $json = json_decode($body, true);
+
+                if ($json === null && json_last_error() !== JSON_ERROR_NONE) {
+                    throw new RuntimeException(sprintf('Server returned non-JSON response: %s', $body), 0, $exception);
+                }
+
+                $errorCode = $json['error']['code'];
+                $errorMessage = $json['error']['message'];
+
+                throw new RuntimeException($errorMessage, $errorCode, $exception);
+            }
+        }
+
+        throw new ConnectException(
+            sprintf('No more servers available, exception from last server: %s', $exception->getMessage()),
+            $exception->getRequest(),
+            $exception
+        );
+    }
+
+    private function send_request(string $server, string $query, array $parameters, bool $bulk_mode = false)
+    {
+        $args_name = 'args';
+        if ($bulk_mode) {
+            $args_name = 'bulk_args';
+        }
+        $options = array_merge($this->httpOptions, [
+            'base_uri' => sprintf('%s://%s', $this->protocol, $server),
+            'json'     => [
+                'stmt' => $query,
+                $args_name => $parameters,
+            ],
+        ]);
+
+        $response     = $this->httpClient->request('POST', '/_sql', $options);
+        $responseBody = json_decode((string)$response->getBody(), true);
+
+        if ($bulk_mode) {
+            return new BulkResponse(
+                $responseBody['results'],
+                $responseBody['cols'],
+                $responseBody['duration']
+            );
+        } else {
+            return new Collection(
+                $responseBody['rows'],
+                $responseBody['cols'],
+                $responseBody['duration'],
+                $responseBody['rowcount']
+            );
+        }
```

Comment:
As outlined at [CrateDB HTTP bulk operations interface](https://crate.io/docs/crate/reference/en/latest/interfaces/http.html#bulk-operations), the response for bulk operations is slightly different from the non-bulk response. I've reflected that by following the white rabbit of typedness.

Another option, which I've actually evaluated a bit beforehand, for easyness, would have been to converge both result formats into the same response object type. In Python DBAPI, where this is only a dictionary [^1], it is not of any concern that this data structure does not have any kind of enforcing schema -- however, in PHP/PDO world, it's differently as it seems both are heavily based on typing in this area. That's why I chose to dedicate a specific response type in PHP now, called `BulkResponse`.

[^1]: https://github.com/crate/crate-python/blob/0.31.1/src/crate/client/cursor.py#L76-L89